File: lt1004.java
// https://leetcode-cn.com/problems/max-consecutive-ones-iii/
// max consecutive ones iii
class Solution {
    public int longestOnes(int[] A, int K) {
        int left = 0, right = 0, curK = 0, len = A.length, maxLen = 0, pre = 0;
        List<Integer> pos = new ArrayList<>();
        while(right < len){
            
            if(A[right] == 0){
                pos.add(right);
                curK++;
            }
            
            if(curK == K + 1){
                maxLen = Math.max(right - left, maxLen);
                left = pos.get(pre++) + 1;
                curK = K;
            }
            
            right++;
        }
        
        maxLen = Math.max(maxLen, right - left);
        
        return maxLen;
    }
}

File: code/lt1004.cpp
#include <vector>
#include <climits>

using namespace std;

class Solution {
public:
    int longestOnes(vector<int>& A, int K) {
        int size = A.size();
        int left = 0, right = 0, max_cnt = INT_MIN, cur_num = 0;
        
        while(right < size)
        {
            if(A[right])
                cur_num++;
            max_cnt = max(cur_num, max_cnt);
            if(right - left + 1 - max_cnt > K)
            {
                if(A[left++])
                    cur_num--;
            }
            right++;
        }

        return max_cnt + K > A.size() ? A.size() : max_cnt + K;
    
    }
};